File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = Integer.MAX_VALUE;
        int[] dp = new int[amount + 1];
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            dp[i] = max;
        }

        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (i < coin) continue;

                if (dp[i - coin] != max) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        return dp[amount] == max ? -1 : dp[amount];
    }
}

File: 0322-coin-change/0322-coin-change.kt
class Solution {
    fun coinChange(coins: IntArray, amount: Int): Int {
        val max = Int.MAX_VALUE
        val dp = IntArray(amount + 1) { max }
        dp[0] = 0

        for (i in 1..amount) {
            for (coin in coins) {
                if (i < coin) continue

                if (dp[i - coin] != max) {
                    dp[i] = minOf(dp[i], dp[i - coin] + 1)
                }
            }
        }

        return if (dp[amount] == max) -1 else dp[amount]
    }
}